File: Data Structures & Algorithms/subsets/submission-1.go
func subsets(nums []int) [][]int {
    var result [][]int
    var subset []int

    var backtrack func(int)
    backtrack = func(start int) {
        result = append(result, append([]int(nil), subset...))

        for i := start; i < len(nums); i++ {
            subset = append(subset, nums[i])
            backtrack(i + 1)

            subset = subset[:len(subset)-1]
        }
    }

    backtrack(0)
    return result
}

File: Data Structures & Algorithms/combination-target-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let result = [];
        let subset = [];
        let sum = 0;

        let resultSet = new Set();

        function backtrack(start) {
            if(sum == target) {
                result.push([...subset])
            }

            for(let i=start; i<nums.length; i++){
                if(sum + nums[i] <= target) {
                    subset.push(nums[i]);
                    sum += nums[i];
                    backtrack(i);
                    sum -= nums[i];
                    subset.pop();
                }
            }
        }

        backtrack(0);
        return result;
    }
}
